### PR TITLE
CFE-745: Rebase cert-manager v1.10

### PR DESCRIPTION
**Pull request by swghosh**
Rebase the operator to use [cert-manager v1.10.2](https://github.com/cert-manager/cert-manager/releases/tag/v1.10.2).

Signed-off-by: Swarup Ghosh <swghosh@redhat.com>

### v1.10.2

> v1.10.2 is primarily a performance enhancement release which might reduce memory consumption by up to 50% in some cases thanks.
> 
> It also patches several vulnerabilities reported by scanners and updates the base images used for cert-manager containers. In addition, it removes a potentially confusing log line which had been introduced in v1.10.0 which implied that an error had occurred when using external issuers even though there'd been no error.
> 
> **Feature**: https://github.com/cert-manager/cert-manager/pull/5647
> **Bugs/Regression**: https://github.com/cert-manager/cert-manager/pull/5704, https://github.com/cert-manager/cert-manager/pull/5676, https://github.com/cert-manager/cert-manager/pull/5687, https://github.com/cert-manager/cert-manager/pull/5592, https://github.com/cert-manager/cert-manager/pull/5635, https://github.com/cert-manager/cert-manager/pull/5620, https://github.com/cert-manager/cert-manager/pull/5682
> **Others**: https://github.com/cert-manager/cert-manager/pull/5677

### v1.10.1

> cert-manager v1.10.1 is a bug fix release which fixes a problem which prevented the Venafi Issuer from connecting to TPP servers where the vedauth API endpoints were configured to accept client certificates.
> It is also compiled with a newer version of Go 1.19 (v1.19.3) which fixes some vulnerabilities in the Go standard library.
> 
> **Bugs/Regression**: https://github.com/cert-manager/cert-manager/pull/5576, https://github.com/cert-manager/cert-manager/pull/5560
> 

### v1.10.0

> Version 1.10 adds a variety of quality-of-life fixes and features including improvements to the test suite.
> 
> **Breaking Changes**: This change is only relevant if you install cert-manager using Helm or the static manifest files. v1.10.0 https://github.com/cert-manager/cert-manager/pull/5410 of containers in pods created by cert-manager. The names are changed to better reflect what they do; for example, the container in the controller pod had its name changed from cert-manager to cert-manager-controller, and the webhook pod had its container name changed from cert-manager to cert-manager-webhook.
> 
> **Feature**: https://github.com/cert-manager/cert-manager/pull/5461, https://github.com/cert-manager/cert-manager/pull/5358, https://github.com/cert-manager/cert-manager/pull/5336, https://github.com/cert-manager/cert-manager/pull/5379, https://github.com/cert-manager/cert-manager/pull/5376, https://github.com/cert-manager/cert-manager/pull/5387, https://github.com/cert-manager/cert-manager/pull/5487, https://github.com/cert-manager/cert-manager/pull/5208
> **Bugs/Regression**:    https://github.com/cert-manager/cert-manager/pull/5323, https://github.com/cert-manager/cert-manager/pull/5339, https://github.com/cert-manager/cert-manager/pull/5259, https://github.com/cert-manager/cert-manager/pull/5366, https://github.com/cert-manager/cert-manager/pull/5022
> **Others**:    https://github.com/cert-manager/cert-manager/pull/5163, https://github.com/cert-manager/cert-manager/pull/5466, https://github.com/cert-manager/cert-manager/pull/5340, https://github.com/cert-manager/cert-manager/pull/5456, https://github.com/cert-manager/cert-manager/pull/5401, https://github.com/cert-manager/cert-manager/pull/5417, https://github.com/cert-manager/cert-manager/pull/5410

### CODE REVIEW

**Comment by swghosh** [comment]
This PR bumps golang from 1.17 to 1.19 and thus openshift/release#35770 is required to be merged for CI to pass on this PR.
Additionally, since the upstream cert-manager images from quay.io/jetstack/cert-manager needs to be mirrored on our CI for this to work, we also need openshift/release#35771.

**Review by thejasn — CHANGES_REQUESTED** [review]
Can you also update the description to contain important release notes from upstream between 1.9.1 to 1.10.2 (just the important ones). This will help later during doc taks.

File: Dockerfile [hunk 1]
```diff
@@ -1,9 +1,9 @@
-FROM docker.io/golang:1.17 AS builder
+FROM docker.io/golang:1.19 AS builder
 WORKDIR /go/src/github.com/openshift/cert-manager-operator
 COPY . .
 RUN make build --warn-undefined-variables
 
-FROM registry.access.redhat.com/ubi9-minimal:9.1.0
+FROM registry.redhat.io/ubi9/ubi-minimal:9.1.0
```
[the remark this answers is not in the document]
Comment (swghosh) [reply Dockerfile]: I am unsure if `registry.access.redhat.com` is publicly pull-able.
Can we change it to: [docker.io/redhat/ubi9-minimal:9.1](https://hub.docker.com/r/redhat/ubi9-minimal) instead? It is a publicly pull-able image with Red Hat as the publisher.

xref: https://developers.redhat.com/products/rhel/ubi

File: bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml [hunk 2]
```diff
@@ -32,14 +32,14 @@ spec:
             - --dynamic-serving-ca-secret-namespace=$(POD_NAMESPACE)
             - --dynamic-serving-ca-secret-name=cert-manager-webhook-ca
             - --dynamic-serving-dns-names=cert-manager-webhook,cert-manager-webhook.cert-manager,cert-manager-webhook.cert-manager.svc
-          command:
-            - /app/cmd/webhook/webhook
+            - --dynamic-serving-dns-names=cert-manager-webhook,cert-manager-webhook.cert-manager,cert-manager-webhook.cert-manager.svc
```
[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml:35]: Okay, checking this.

[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml:35]: Yeah, we have a problem here.
The args is actually specified thrice in upstream which is supposedly causing the duplication issue.

(ref: https://github.com/cert-manager/cert-manager/releases/download/v1.10.2/cert-manager.yaml, lines: 5385-5387)

```
---
          - --dynamic-serving-dns-names=cert-manager-webhook
          - --dynamic-serving-dns-names=cert-manager-webhook.$(POD_NAMESPACE)
          - --dynamic-serving-dns-names=cert-manager-webhook.$(POD_NAMESPACE).svc
---
```

[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml:35]: well, jsonnet doing the wonder:
https://github.com/openshift/cert-manager-operator/blob/ef377e45229fe681dc949f99704a6bfd0d940d5a/jsonnet/main.jsonnet#L40
can change/fix it.

[the remark this answers is not in the document]
Comment (thejasn) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml:35]: Looks like upstream pushed the [fix](https://github.com/cert-manager/cert-manager/pull/5163). Can you correct the jsonnet script? (ensure this is in a separate commit)

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply Dockerfile]: Cool, so will update to use `FROM registry.access.redhat.com/ubi9-minimal:9.1.0` as is.

File: bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml:35]: Yes, certainly.

File: bundle/manifests/cert-manager-operator.clusterserviceversion.yaml [hunk 3]
```diff
@@ -711,10 +711,10 @@ spec:
...
+  - image: quay.io/jetstack/cert-manager-webhook:v1.10.2
     name: cert-manager-webhook
-  - image: quay.io/jetstack/cert-manager-cainjector:v1.9.1
+  - image: quay.io/jetstack/cert-manager-cainjector:v1.10.2
     name: cert-manager-ca-injector
-  - image: quay.io/jetstack/cert-manager-controller:v1.9.1
+  - image: quay.io/jetstack/cert-manager-controller:v1.10.2
     name: cert-manager-controller
-  version: 0.0.1
+  version: 1.10.2
```
Comment (thejasn) [inline bundle/manifests/cert-manager-operator.clusterserviceversion.yaml:720]: Change this back to `0.0.1` sine the bundle version was reverted as well.

File: go.mod [hunk 4]
```diff
@@ -1,6 +1,6 @@
 module github.com/openshift/cert-manager-operator
 
-go 1.17
+go 1.19
```
[the remark this answers is not in the document]
Comment (swghosh) [reply go.mod:3]: Supposedly, go version specified in go.mod cannot include minor version of golang. [[1]](https://go.dev/doc/modules/gomod-ref#go-syntax).
However, I can confirm that the golang binaries we use in the corresponding container base images, have available `go> =1.19.2`.

 1. docker.io/golang:1.19 in Dockerfile
```
$ podman run docker.io/golang:1.19 go version
go version go1.19.5 linux/amd64
```
2. registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13 in Makefile
```
$ podman run registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13 go version
go version go1.19.2 linux/amd64
```
3. registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13 in .ci-operator.yaml
```
$ podman run registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13 go version
go version go1.19.2 linux/amd64
```

[1] https://go.dev/doc/modules/gomod-ref#go-syntax

File: bundle/manifests/cert-manager-operator.clusterserviceversion.yaml (hunk 3 above)
Comment (swghosh) [reply bundle/manifests/cert-manager-operator.clusterserviceversion.yaml]: Done.

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply Dockerfile]: Updated.

File: bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml]: Updated the jsonnet accordingly, PTAL at https://github.com/openshift/cert-manager-operator/pull/98/commits/04e6ac093bf616fee4a48a2009d5645d4b4f4186.

[the remark this answers is not in the document]
Comment (thejasn) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml]: Hmm, taking a look at upstream changes and the jsonnet script looks like we don't need to process args at all and upstream provides args exactly like how we want it.

[the remark this answers is not in the document]
Comment (swghosh) [reply bindata/cert-manager-deployment/webhook/cert-manager-webhook-deployment.yaml]: Updated as per our last discussion to use a rule that combines values of all `--dynamic-serving-dns-names` into a single comma separated arg.

File: bindata/cert-manager-deployment/controller/cert-manager-deployment.yaml [hunk 5]
```diff
@@ -27,31 +27,34 @@ spec:
...
         app.kubernetes.io/name: cert-manager
-        app.kubernetes.io/version: v1.9.1
+        app.kubernetes.io/version: v1.10.2
     spec:
       containers:
         - args:
             - --v=2
             - --cluster-resource-namespace=$(POD_NAMESPACE)
             - --leader-election-namespace=kube-system
-          command:
```
Comment (thejasn) [inline bindata/cert-manager-deployment/controller/cert-manager-deployment.yaml:37]: As per discussion today, lets add this logic back.

File: .ci-operator.yaml [hunk 6]
```diff
@@ -1,4 +1,4 @@
 build_root_image:
   name: release
   namespace: openshift
-  tag: rhel-8-release-golang-1.17-openshift-4.10
+  tag: golang-1.19
```
Comment (thejasn) [inline .ci-operator.yaml:4]: Can you change this to `rhel-9-golang-1.19-openshift-4.13` ?

File: bindata/cert-manager-deployment/controller/cert-manager-deployment.yaml (hunk 5 above)
Comment (swghosh) [reply bindata/cert-manager-deployment/controller/cert-manager-deployment.yaml]: Added in https://github.com/openshift/cert-manager-operator/pull/98/commits/f381426fe510f7db95adec37253bf6d2dad6d970.

File: .ci-operator.yaml (hunk 6 above)
Comment (swghosh) [reply .ci-operator.yaml:4]: `rhel-9-golang-1.19-openshift-4.13` tag isn't available yet, which is the same reason why I used just `golang-1.19`.

Comment (swghosh) [reply .ci-operator.yaml:4]: JFYR:

```
$ skopeo list-tags docker://registry.ci.openshift.org/openshift/release
{
    "Repository": "registry.ci.openshift.org/openshift/release",
    "Tags": [
        "tectonic-console-builder-v24",
        "golang-1.18",
        "rhel-7-release-openshift-4.11",
        "tectonic-console-builder-v21",
        "rhel-8-release-golang-1.19-openshift-4.13",
        "jdk-8",
        "rhel-7-release-openshift-4.6",
        "rhel-7-release-openshift-4.7",
        "rhel-8-release-golang-1.15-openshift-4.7",
        "rhel-8-release-golang-1.18-openshift-4.12",
        "tectonic-console-builder-v23",
        "golang-1.19",
        "golang-1.16-vault-typescript",
        "golang-1.17",
        "rhel-golang-1.12",
        "tectonic-console-builder-v20",
        "golang-1.16",
        "rhel-8-release-golang-1.19-openshift-4.12",
        "rhel-7-release-openshift-4.6.art",
        "nodejs-8-browser-tests",
        "rhel-7-release-openshift-4.10",
        "rhel-8-release-golang-1.16-openshift-4.10",
        "tectonic-console-builder-v17",
        "golang-1.11",
        "rhel-8-release-golang-1.16-openshift-4.8",
        "rhel-8-release-openshift-4.6",
        "tectonic-console-builder-v22",
        "rhel-7-release-openshift-4.9",
        "rhel-7-golang-1.11",
        "golang-1.14",
        "golang-1.16-vault",
        "rhel-8-release-golang-1.14-openshift-4.7",
        "rhel-8-release-golang-1.15-openshift-4.8",
        "tectonic-console-builder-v18",
        "tectonic-console-builder-v19",
        "golang-1.15",
        "rhel-8-release-golang-1.18-openshift-4.11",
        "ruby-25",
        "rhel-8-release-golang-1.17-openshift-4.10",
        "rhel-7-release-openshift-4.8.art",
        "golang-1.9",
        "golang-1.8",
        "rhel-8-release-golang-1.14-openshift-4.8",
        "rhel-8-release-openshift-4.3",
        "golang-1.10",
        "nodejs-8",
        "rhel-7-golang-1.12",
        "rhel-7-release-openshift-4.7.art",
        "rhel-8-release-golang-1.14-openshift-4.6",
        "rhel-8-release-golang-1.17-openshift-4.11",
        "rhel-golang-1.11",
        "golang-1.12",
        "rhel-7-release-openshift-4.8",
        "rhel-8-release-golang-1.15-openshift-4.6",
        "rhel-8-release-golang-1.16-openshift-4.9",
        "golang-1.13"
    ]
}
```

Comment (swghosh) [reply .ci-operator.yaml:4]: alternative (apart from this) is to use: `registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13` similar to what is in: [o/oc-mirror:.ci-operator.yaml](https://github.com/openshift/oc-mirror/blob/main/.ci-operator.yaml).

**Comment by xingxingxia** [comment]
Tested with cluster-bot env of:
```
launch openshift/cert-manager-operator#98
```
Basic issuing functions work well. Adding label:
/label qe-approved

File: .ci-operator.yaml (hunk 6 above)
Comment (thejasn) [reply .ci-operator.yaml:4]:
```
podman pull registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13
Trying to pull registry.ci.openshift.org/ocp/builder:rhel-9-golang-1.19-openshift-4.13...
Getting image source signatures
Copying blob fda3ab9c84ea [=============>------------------------] 54.9MiB / 150.1MiB
Copying blob 793a8691df4a [========================>-------------] 49.4MiB / 75.2MiB
Copying blob 70e0ef827acf [==>-----------------------------------] 46.8MiB / 574.5MiB
```
Seems to be available. 

Also update corresponding pull specs in release.yaml,
- https://github.com/openshift/release/blob/7b17bfb10eb58bf18cef1ccd877f4bc39d8f6ee0/ci-operator/config/openshift/cert-manager-operator/openshift-cert-manager-operator-master.yaml#L18
- https://github.com/openshift/release/blob/7b17bfb10eb58bf18cef1ccd877f4bc39d8f6ee0/ci-operator/config/openshift/cert-manager-operator/openshift-cert-manager-operator-master.yaml#L29

Comment (swghosh) [reply .ci-operator.yaml:4]: yep, so changes from https://github.com/openshift/release/pull/35770 might have got reverted as our repo's master branch still got old tags.

As per discussion, changing the operator's .ci-operator.yaml, ci Dockerfile in separate PR: #100 which should be preceded for sync by https://github.com/openshift/release/pull/35870.

**Comment by thejasn** [comment]
/cc @bergerhoffer @davemulford 
We will be bumping `cert-manager` to `v1.10.2` since `v1.9.x` has reached EOL.

**Comment by swghosh** [comment]
/remove-hold
#100 was merged

**Pull request merged**